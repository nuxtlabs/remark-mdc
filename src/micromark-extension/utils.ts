import type { Effects } from 'micromark-factory-space'

// Measure the number of character codes in chunks.
// Counts tabs based on their expanded size, and CR+LF as one character.
export function sizeChunks (chunks: any[]) {
  let index = -1
  let size = 0

  while (++index < chunks.length) {
    size += typeof chunks[index] === 'string' ? chunks[index].length : 1
  }

  return size
}

export function prefixSize (events: any, type: string) {
  const tail = events[events.length - 1]
  if (!tail || tail[1].type !== type) { return 0 }
  return sizeChunks(tail[2].sliceStream(tail[1]))
}

/**
 * Calculate line indention size, line indention could be consists of multiple `linePrefix` events
 * @param events parser tokens
 * @returns line indention size
 */
export function linePrefixSize (events: any[]) {
  let size = 0
  let index = events.length - 1
  let tail = events[index]
  while (index >= 0 && tail && tail[1].type === 'linePrefix' && tail[0] === 'exit') {
    size += sizeChunks(tail[2].sliceStream(tail[1]))
    index -= 1
    tail = events[index]
  }

  return size
}

/**
 * Manage token state
 */
export const useTokenState = (tokenName: string) => {
  const token = {
    isOpen: false,
    /**
     * Enter into token, close previous open token if any
     */
    enter: (effects: Effects) => {
      const initialState = token.isOpen
      token.exit(effects)
      effects.enter(tokenName)
      token.isOpen = true

      // Revert to initial state
      return () => {
        token.isOpen = initialState
      }
    },
    /**
     * Enter into token only once, if token is already open, do nothing
     */
    enterOnce: (effects: Effects) => {
      const initialState = token.isOpen

      if (!token.isOpen) {
        effects.enter(tokenName)
        token.isOpen = true
      }

      // Revert to initial state
      return () => {
        token.isOpen = initialState
      }
    },
    /**
     * Exit from token if it is open
     */
    exit: (effects: Effects) => {
      const initialState = token.isOpen
      if (token.isOpen) {
        effects.exit(tokenName)
        token.isOpen = false
      }

      // Revert to initial state
      return () => {
        token.isOpen = initialState
      }
    }
  }
  return token
}
