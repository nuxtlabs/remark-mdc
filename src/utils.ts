export const NON_UNWRAPABLE_TYPES = [
  'componentContainerSection',
  'componentContainerDataSection',
  'containerComponent',
  'leafComponent'
]
