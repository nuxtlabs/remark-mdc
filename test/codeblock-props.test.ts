import { expect, describe } from 'vitest'
import { runMarkdownTests } from './utils'

describe('codeblock-props', () => {
  runMarkdownTests({
    'YamlProps': {
      mdcOptions: {
        experimental: {
          componentCodeBlockYamlProps: true,
        },
      },
      markdown: '::with-frontmatter-yaml\n```yaml [props]\narray:\n  - item\n  - itemKey: value\nkey: value\n```\n::',
      extra(_markdown, ast, _expected) {
        expect(ast.children[0].type).toEqual('containerComponent')
        expect(ast.children[0].fmAttributes).toEqual({ array: ['item', { itemKey: 'value' }], key: 'value' })
      },
    },
    'notYamlProps': {
      mdcOptions: {
        experimental: {
          componentCodeBlockYamlProps: true,
        },
      },
      markdown: '::with-frontmatter-yaml\n```yaml\nkey: value\narray:\n  - item\n  - itemKey: value\n```\n::',
      extra(_markdown, ast, _expected) {
        expect(ast.children[0].type).toEqual('containerComponent')
        expect(ast.children[0].fmAttributes).toEqual({})
      },
    },
    'notYamlProps2': {
      mdcOptions: {
        experimental: {
          componentCodeBlockYamlProps: true,
        },
      },
      markdown: '::with-frontmatter-yaml\n```yaml[props]\nkey: value\narray:\n  - item\n  - itemKey: value\n```\n::',
    },
    'notYamlProps3': {
      mdcOptions: {
        experimental: {
          componentCodeBlockYamlProps: true,
        },
      },
      markdown: '::with-frontmatter-yaml\n```yaml [yaml]\nkey: value\narray:\n  - item\n  - itemKey: value\n```\n::',
    },
    'shouldConvertYamlProps': {
      mdcOptions: {
        experimental: {
          componentCodeBlockYamlProps: true,
        },
      },
      markdown: '::with-frontmatter-yaml\n---\narray:\n  - item\n  - itemKey: value\nkey: value\n---\n::',
      expected: '::with-frontmatter-yaml\n```yaml [props]\narray:\n  - item\n  - itemKey: value\nkey: value\n```\n::',
    },
    // shouldConvertYamlPropsWithoutOption: {
    //   markdown: '::with-frontmatter-yaml\n```yaml [props]\narray:\n  - item\n  - itemKey: value\nkey: value\n```\n::',
    //   expected: '::with-frontmatter-yaml\n---\narray:\n  - item\n  - itemKey: value\nkey: value\n---\n::'
    // },
    'yamlProps1': {
      mdcOptions: {
        experimental: {
          componentCodeBlockYamlProps: true,
        },
      },
      markdown: [
        '::with-frontmatter-yaml1',
        '```yaml [props]',
        'key: value',
        'key2:',
        '  subkey: value',
        '  subkey2: value',
        'array:',
        '  - item',
        '  - itemKey: value',
        '```',
        '::',
      ].join('\n'),
      expected: [
        '::with-frontmatter-yaml1',
        '```yaml [props]',
        'array:',
        '  - item',
        '  - itemKey: value',
        'key: value',
        'key2:',
        '  subkey: value',
        '  subkey2: value',
        '```',
        '::',
      ].join('\n'),
    },
    'nested-component-yamlProps': {
      mdcOptions: {
        experimental: {
          componentCodeBlockYamlProps: true,
        },
      },
      markdown: [
        '::with-frontmatter-and-nested-component-yaml',
        '```yaml [props]',
        'array:',
        '  - item',
        '  - itemKey: value',
        'key: value',
        '```',
        'Default slot',
        '',
        '#secondary-slot',
        'Secondary slot value',
        '',
        '  :::hello',
        '  ```yaml [props]',
        '  key: value',
        '  ```',
        '  :::',
        '::',
      ].join('\n'),
      extra(_markdown, ast, _expected) {
        expect(ast.children[0].type).toEqual('containerComponent')
        expect(ast.children[0].fmAttributes).toEqual({ array: ['item', { itemKey: 'value' }], key: 'value' })
        expect(ast.children[0].children[1].children[1].fmAttributes).toEqual({ key: 'value' })
      },
    },
  })
})
