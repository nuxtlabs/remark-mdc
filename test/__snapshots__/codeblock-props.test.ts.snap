// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`codeblock-props > YamlProps 1`] = `
{
  "children": [
    {
      "attributes": {},
      "children": [],
      "data": {
        "hName": "with-frontmatter-yaml",
        "hProperties": {
          ":array": "["item",{"itemKey":"value"}]",
          "key": "value",
        },
      },
      "fmAttributes": {
        "array": [
          "item",
          {
            "itemKey": "value",
          },
        ],
        "key": "value",
      },
      "mdc": {
        "codeBlockProps": true,
      },
      "name": "with-frontmatter-yaml",
      "position": {
        "end": {
          "column": 3,
          "line": 8,
          "offset": 92,
        },
        "start": {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "rawData": "array:
  - item
  - itemKey: value
key: value",
      "type": "containerComponent",
    },
  ],
  "position": {
    "end": {
      "column": 3,
      "line": 8,
      "offset": 92,
    },
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`codeblock-props > YamlProps 2`] = `
{
  "children": [
    {
      "attributes": {},
      "children": [],
      "data": {
        "hName": "with-frontmatter-yaml",
        "hProperties": {
          ":array": "["item",{"itemKey":"value"}]",
          "key": "value",
        },
      },
      "fmAttributes": {
        "array": [
          "item",
          {
            "itemKey": "value",
          },
        ],
        "key": "value",
      },
      "mdc": {
        "codeBlockProps": true,
      },
      "name": "with-frontmatter-yaml",
      "position": {
        "end": {
          "column": 3,
          "line": 8,
          "offset": 92,
        },
        "start": {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "rawData": "array:
  - item
  - itemKey: value
key: value",
      "type": "containerComponent",
    },
  ],
  "position": {
    "end": {
      "column": 3,
      "line": 8,
      "offset": 92,
    },
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`codeblock-props > nested-component-yamlProps 1`] = `
{
  "children": [
    {
      "attributes": {},
      "children": [
        {
          "children": [
            {
              "position": {
                "end": {
                  "column": 13,
                  "line": 8,
                  "offset": 123,
                },
                "start": {
                  "column": 1,
                  "line": 8,
                  "offset": 111,
                },
              },
              "type": "text",
              "value": "Default slot",
            },
          ],
          "position": {
            "end": {
              "column": 13,
              "line": 8,
              "offset": 123,
            },
            "start": {
              "column": 1,
              "line": 8,
              "offset": 111,
            },
          },
          "type": "paragraph",
        },
        {
          "attributes": {},
          "children": [
            {
              "children": [
                {
                  "position": {
                    "end": {
                      "column": 21,
                      "line": 11,
                      "offset": 161,
                    },
                    "start": {
                      "column": 1,
                      "line": 11,
                      "offset": 141,
                    },
                  },
                  "type": "text",
                  "value": "Secondary slot value",
                },
              ],
              "position": {
                "end": {
                  "column": 21,
                  "line": 11,
                  "offset": 161,
                },
                "start": {
                  "column": 1,
                  "line": 11,
                  "offset": 141,
                },
              },
              "type": "paragraph",
            },
            {
              "attributes": {},
              "children": [],
              "data": {
                "hName": "hello",
                "hProperties": {
                  "key": "value",
                },
              },
              "fmAttributes": {
                "key": "value",
              },
              "mdc": {
                "codeBlockProps": true,
              },
              "name": "hello",
              "position": {
                "end": {
                  "column": 4,
                  "line": 17,
                  "offset": 206,
                },
                "start": {
                  "column": 1,
                  "line": 13,
                  "offset": 163,
                },
              },
              "rawData": "key: value",
              "type": "containerComponent",
            },
          ],
          "data": {
            "hName": "component-slot",
            "hProperties": {
              "v-slot:secondary-slot": "",
            },
          },
          "name": "secondary-slot",
          "position": {
            "end": {
              "column": 3,
              "line": 18,
              "offset": 209,
            },
            "start": {
              "column": 2,
              "line": 10,
              "offset": 126,
            },
          },
          "type": "componentContainerSection",
        },
      ],
      "data": {
        "hName": "with-frontmatter-and-nested-component-yaml",
        "hProperties": {
          ":array": "["item",{"itemKey":"value"}]",
          "key": "value",
        },
      },
      "fmAttributes": {
        "array": [
          "item",
          {
            "itemKey": "value",
          },
        ],
        "key": "value",
      },
      "mdc": {
        "codeBlockProps": true,
      },
      "name": "with-frontmatter-and-nested-component-yaml",
      "position": {
        "end": {
          "column": 3,
          "line": 18,
          "offset": 209,
        },
        "start": {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "rawData": "array:
  - item
  - itemKey: value
key: value",
      "type": "containerComponent",
    },
  ],
  "position": {
    "end": {
      "column": 3,
      "line": 18,
      "offset": 209,
    },
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`codeblock-props > nested-component-yamlProps 2`] = `
{
  "children": [
    {
      "attributes": {},
      "children": [
        {
          "children": [
            {
              "position": {
                "end": {
                  "column": 13,
                  "line": 8,
                  "offset": 123,
                },
                "start": {
                  "column": 1,
                  "line": 8,
                  "offset": 111,
                },
              },
              "type": "text",
              "value": "Default slot",
            },
          ],
          "position": {
            "end": {
              "column": 13,
              "line": 8,
              "offset": 123,
            },
            "start": {
              "column": 1,
              "line": 8,
              "offset": 111,
            },
          },
          "type": "paragraph",
        },
        {
          "attributes": {},
          "children": [
            {
              "children": [
                {
                  "position": {
                    "end": {
                      "column": 21,
                      "line": 11,
                      "offset": 161,
                    },
                    "start": {
                      "column": 1,
                      "line": 11,
                      "offset": 141,
                    },
                  },
                  "type": "text",
                  "value": "Secondary slot value",
                },
              ],
              "position": {
                "end": {
                  "column": 21,
                  "line": 11,
                  "offset": 161,
                },
                "start": {
                  "column": 1,
                  "line": 11,
                  "offset": 141,
                },
              },
              "type": "paragraph",
            },
            {
              "attributes": {},
              "children": [],
              "data": {
                "hName": "hello",
                "hProperties": {
                  "key": "value",
                },
              },
              "fmAttributes": {
                "key": "value",
              },
              "mdc": {
                "codeBlockProps": true,
              },
              "name": "hello",
              "position": {
                "end": {
                  "column": 4,
                  "line": 17,
                  "offset": 206,
                },
                "start": {
                  "column": 1,
                  "line": 13,
                  "offset": 163,
                },
              },
              "rawData": "key: value",
              "type": "containerComponent",
            },
          ],
          "data": {
            "hName": "component-slot",
            "hProperties": {
              "v-slot:secondary-slot": "",
            },
          },
          "name": "secondary-slot",
          "position": {
            "end": {
              "column": 3,
              "line": 18,
              "offset": 209,
            },
            "start": {
              "column": 2,
              "line": 10,
              "offset": 126,
            },
          },
          "type": "componentContainerSection",
        },
      ],
      "data": {
        "hName": "with-frontmatter-and-nested-component-yaml",
        "hProperties": {
          ":array": "["item",{"itemKey":"value"}]",
          "key": "value",
        },
      },
      "fmAttributes": {
        "array": [
          "item",
          {
            "itemKey": "value",
          },
        ],
        "key": "value",
      },
      "mdc": {
        "codeBlockProps": true,
      },
      "name": "with-frontmatter-and-nested-component-yaml",
      "position": {
        "end": {
          "column": 3,
          "line": 18,
          "offset": 209,
        },
        "start": {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "rawData": "array:
  - item
  - itemKey: value
key: value",
      "type": "containerComponent",
    },
  ],
  "position": {
    "end": {
      "column": 3,
      "line": 18,
      "offset": 209,
    },
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`codeblock-props > notYamlProps 1`] = `
{
  "children": [
    {
      "attributes": {},
      "children": [
        {
          "lang": "yaml",
          "meta": null,
          "position": {
            "end": {
              "column": 4,
              "line": 7,
              "offset": 81,
            },
            "start": {
              "column": 1,
              "line": 2,
              "offset": 24,
            },
          },
          "type": "code",
          "value": "key: value
array:
  - item
  - itemKey: value",
        },
      ],
      "data": {
        "hName": "with-frontmatter-yaml",
        "hProperties": {},
      },
      "fmAttributes": {},
      "name": "with-frontmatter-yaml",
      "position": {
        "end": {
          "column": 3,
          "line": 8,
          "offset": 84,
        },
        "start": {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "containerComponent",
    },
  ],
  "position": {
    "end": {
      "column": 3,
      "line": 8,
      "offset": 84,
    },
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`codeblock-props > notYamlProps 2`] = `
{
  "children": [
    {
      "attributes": {},
      "children": [
        {
          "lang": "yaml",
          "meta": null,
          "position": {
            "end": {
              "column": 6,
              "line": 7,
              "offset": 93,
            },
            "start": {
              "column": 3,
              "line": 2,
              "offset": 26,
            },
          },
          "type": "code",
          "value": "key: value
array:
  - item
  - itemKey: value",
        },
      ],
      "data": {
        "hName": "with-frontmatter-yaml",
        "hProperties": {},
      },
      "fmAttributes": {},
      "name": "with-frontmatter-yaml",
      "position": {
        "end": {
          "column": 3,
          "line": 8,
          "offset": 96,
        },
        "start": {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "containerComponent",
    },
  ],
  "position": {
    "end": {
      "column": 3,
      "line": 8,
      "offset": 96,
    },
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`codeblock-props > notYamlProps2 1`] = `
{
  "children": [
    {
      "attributes": {},
      "children": [
        {
          "lang": "yaml[props]",
          "meta": null,
          "position": {
            "end": {
              "column": 4,
              "line": 7,
              "offset": 88,
            },
            "start": {
              "column": 1,
              "line": 2,
              "offset": 24,
            },
          },
          "type": "code",
          "value": "key: value
array:
  - item
  - itemKey: value",
        },
      ],
      "data": {
        "hName": "with-frontmatter-yaml",
        "hProperties": {},
      },
      "fmAttributes": {},
      "name": "with-frontmatter-yaml",
      "position": {
        "end": {
          "column": 3,
          "line": 8,
          "offset": 91,
        },
        "start": {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "containerComponent",
    },
  ],
  "position": {
    "end": {
      "column": 3,
      "line": 8,
      "offset": 91,
    },
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`codeblock-props > notYamlProps2 2`] = `
{
  "children": [
    {
      "attributes": {},
      "children": [
        {
          "lang": "yaml[props]",
          "meta": null,
          "position": {
            "end": {
              "column": 6,
              "line": 7,
              "offset": 100,
            },
            "start": {
              "column": 3,
              "line": 2,
              "offset": 26,
            },
          },
          "type": "code",
          "value": "key: value
array:
  - item
  - itemKey: value",
        },
      ],
      "data": {
        "hName": "with-frontmatter-yaml",
        "hProperties": {},
      },
      "fmAttributes": {},
      "name": "with-frontmatter-yaml",
      "position": {
        "end": {
          "column": 3,
          "line": 8,
          "offset": 103,
        },
        "start": {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "containerComponent",
    },
  ],
  "position": {
    "end": {
      "column": 3,
      "line": 8,
      "offset": 103,
    },
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`codeblock-props > notYamlProps3 1`] = `
{
  "children": [
    {
      "attributes": {},
      "children": [
        {
          "lang": "yaml",
          "meta": "[yaml]",
          "position": {
            "end": {
              "column": 4,
              "line": 7,
              "offset": 88,
            },
            "start": {
              "column": 1,
              "line": 2,
              "offset": 24,
            },
          },
          "type": "code",
          "value": "key: value
array:
  - item
  - itemKey: value",
        },
      ],
      "data": {
        "hName": "with-frontmatter-yaml",
        "hProperties": {},
      },
      "fmAttributes": {},
      "name": "with-frontmatter-yaml",
      "position": {
        "end": {
          "column": 3,
          "line": 8,
          "offset": 91,
        },
        "start": {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "containerComponent",
    },
  ],
  "position": {
    "end": {
      "column": 3,
      "line": 8,
      "offset": 91,
    },
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`codeblock-props > notYamlProps3 2`] = `
{
  "children": [
    {
      "attributes": {},
      "children": [
        {
          "lang": "yaml",
          "meta": "[yaml]",
          "position": {
            "end": {
              "column": 6,
              "line": 7,
              "offset": 100,
            },
            "start": {
              "column": 3,
              "line": 2,
              "offset": 26,
            },
          },
          "type": "code",
          "value": "key: value
array:
  - item
  - itemKey: value",
        },
      ],
      "data": {
        "hName": "with-frontmatter-yaml",
        "hProperties": {},
      },
      "fmAttributes": {},
      "name": "with-frontmatter-yaml",
      "position": {
        "end": {
          "column": 3,
          "line": 8,
          "offset": 103,
        },
        "start": {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "containerComponent",
    },
  ],
  "position": {
    "end": {
      "column": 3,
      "line": 8,
      "offset": 103,
    },
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`codeblock-props > shouldConvertYamlProps 1`] = `
{
  "children": [
    {
      "attributes": {},
      "children": [],
      "data": {
        "hName": "with-frontmatter-yaml",
        "hProperties": {
          ":array": "["item",{"itemKey":"value"}]",
          "key": "value",
        },
      },
      "fmAttributes": {
        "array": [
          "item",
          {
            "itemKey": "value",
          },
        ],
        "key": "value",
      },
      "name": "with-frontmatter-yaml",
      "position": {
        "end": {
          "column": 3,
          "line": 8,
          "offset": 80,
        },
        "start": {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "rawData": "
array:
  - item
  - itemKey: value
key: value
---",
      "type": "containerComponent",
    },
  ],
  "position": {
    "end": {
      "column": 3,
      "line": 8,
      "offset": 80,
    },
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`codeblock-props > shouldConvertYamlProps 2`] = `
{
  "children": [
    {
      "attributes": {},
      "children": [],
      "data": {
        "hName": "with-frontmatter-yaml",
        "hProperties": {
          ":array": "["item",{"itemKey":"value"}]",
          "key": "value",
        },
      },
      "fmAttributes": {
        "array": [
          "item",
          {
            "itemKey": "value",
          },
        ],
        "key": "value",
      },
      "name": "with-frontmatter-yaml",
      "position": {
        "end": {
          "column": 3,
          "line": 8,
          "offset": 80,
        },
        "start": {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "rawData": "
array:
  - item
  - itemKey: value
key: value
---",
      "type": "containerComponent",
    },
  ],
  "position": {
    "end": {
      "column": 3,
      "line": 8,
      "offset": 80,
    },
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`codeblock-props > yamlProps1 1`] = `
{
  "children": [
    {
      "attributes": {},
      "children": [],
      "data": {
        "hName": "with-frontmatter-yaml1",
        "hProperties": {
          ":array": "["item",{"itemKey":"value"}]",
          ":key2": "{"subkey":"value","subkey2":"value"}",
          "key": "value",
        },
      },
      "fmAttributes": {
        "array": [
          "item",
          {
            "itemKey": "value",
          },
        ],
        "key": "value",
        "key2": {
          "subkey": "value",
          "subkey2": "value",
        },
      },
      "mdc": {
        "codeBlockProps": true,
      },
      "name": "with-frontmatter-yaml1",
      "position": {
        "end": {
          "column": 3,
          "line": 11,
          "offset": 132,
        },
        "start": {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "rawData": "key: value
key2:
  subkey: value
  subkey2: value
array:
  - item
  - itemKey: value",
      "type": "containerComponent",
    },
  ],
  "position": {
    "end": {
      "column": 3,
      "line": 11,
      "offset": 132,
    },
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`codeblock-props > yamlProps1 2`] = `
{
  "children": [
    {
      "attributes": {},
      "children": [],
      "data": {
        "hName": "with-frontmatter-yaml1",
        "hProperties": {
          ":array": "["item",{"itemKey":"value"}]",
          ":key2": "{"subkey":"value","subkey2":"value"}",
          "key": "value",
        },
      },
      "fmAttributes": {
        "array": [
          "item",
          {
            "itemKey": "value",
          },
        ],
        "key": "value",
        "key2": {
          "subkey": "value",
          "subkey2": "value",
        },
      },
      "mdc": {
        "codeBlockProps": true,
      },
      "name": "with-frontmatter-yaml1",
      "position": {
        "end": {
          "column": 3,
          "line": 11,
          "offset": 132,
        },
        "start": {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "rawData": "key: value
key2:
  subkey: value
  subkey2: value
array:
  - item
  - itemKey: value",
      "type": "containerComponent",
    },
  ],
  "position": {
    "end": {
      "column": 3,
      "line": 11,
      "offset": 132,
    },
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;
